rfcs/text/0000-create-file-mach-o.md

```diff
@@ -0,0 +1,155 @@
+# 0000: Create the Mach-O sub-field of the File fieldset
+
+- Stage: **0 (strawperson)**
+- Date: **TBD**
+
+Create the Mach Object (Mach-O) sub-field, of the `file` top-level fieldset. This document metadata can be used for malware research, as well as coding and other application development efforts.
+
+## Fields
+
+**Stage 0**
+
+This RFC is to create the Mach-O sub-field within the `file.` fieldset. This will include 35 sub-fields.
+
+|   Name                                     |   Type     |   Description                                                               |
+|--------------------------------------------|------------|-----------------------------------------------------------------------------|
+|   file.mach-o.cpu                          |   object   |   CPU information for the file.                                             |
+|   file.mach-o.cpu.architecture             |   keyword  |   CPU architecture target for the file.                                     |
+|   file.mach-o.cpu.byte_order               |   keyword  |   CPU byte order for the file.                                              |
+|   file.mach-o.cpu.subtype                  |   keyword  |   CPU subtype for the file.                                                 |
+|   file.mach-o.cpu.type                     |   keyword  |   CPU type for the file.                                                    |
+|   file.mach-o.headers                      |   object   |   Header information for the file.                                          |
+|   file.mach-o.headers.commands             |   object   |   Header load commands for the file.                                        |
+|   file.mach-o.headers.commands.number      |   long     |   Number of load commands for the Mach-O header.                            |
+|   file.mach-o.headers.commands.size        |   long     |   Size of load commands of the Mach-O header.                               |
+|   file.mach-o.headers.commands.type        |   keyword  |   Type of the load commands for the Mach-O header.                          |
+|   file.mach-o.headers.magic                |   keyword  |   Magic field of the Mach-O header.                                         |
+|   file.mach-o.headers.flags                |   keyword  |   Flags set in the Mach-O header.                                           |
+|   file.mach-o.segments                     |   object   |   Segment information for the file.                                         |
```

Comment:
I would think that the same `entropy` and hash calculations as PE fields would make sense here--also import/export tables.

Reply:
I apologize, I don't have entropy or import/export tables data in mach-o.

What am I missing?

Reply:
Just pinging on this @andrewstucki

Reply:
I'm fine with not including this in the first pass, but I could have sworn that at least VT contained entropy calculations for either segments or sections.

If they're also missing import/export tables in the samples you're working with, I'm fine punting on that until we have a use-case to add them in.

I'm going to try and get around to having some space time and working on the object file parser that I started for beats quite some time back, so if we get this as a foundation, maybe I can extend the schema as needed.

Reply:
also, think rather than `object`, you want this to be `nested`, no?

Reply:
Sorry, I thought that `type: object` was meaning that there were `nested` fields? Is it something different?

Reply:
Should it be changed to `type: nested`?

Reply:
`type: nested` is only necessary if this is intended to use the Elasticsearch nested type data type, but not if the fields are just nested underneath an ECS fieldset (e.g., `macho.segments.name`)

Reply:
in the case of segments, we're dealing with multiple segments, not just a single segment with a given name (i.e. no, `segments.name`, instead `segments[0].name`), so it's either going to need to be `nested` or `flattened` to be queryable. Since not all of the attributes (i.e. offsets) are text, `nested` is the way to go.

Reply:
Done.